File: scripts/algorithms/T/Third Maximum Number/Third Maximum Number.java
// Runtime: 0 ms (Top 100.0%) | Memory: 42.70 MB (Top 89.77%)

class Solution {
    public int thirdMax(int[] nums) {
	
		// taking long data type since array can contain Integer.MIN_VALUE
        long max = Long.MIN_VALUE, sMax = Long.MIN_VALUE, tMax = Long.MIN_VALUE;
        for (int i : nums) {
            if (i > max) {
                tMax = sMax;
                sMax = max;
                max = i;
            } else if (i > sMax && i != max) {
                tMax = sMax;
                sMax = i;
            } else if (i > tMax && i != sMax && i != max) {
                tMax = i;
            }
        }
		
		// if thirdMax was not updated, return the first Max
		
        return tMax == Long.MIN_VALUE ? (int) max : (int) tMax;
    }
}

File: scripts/algorithms/T/Third Maximum Number/Third Maximum Number.cpp
// Runtime: 12 ms (Top 11.87%) | Memory: 11.00 MB (Top 18.16%)

class Solution {
public:
 int thirdMax(vector<int>& nums) {
           set<int>s;
        for(int i=0;i<nums.size();i++){
            s.insert(nums[i]);
        }
        if(s.size()>=3){   // when set size >=3 means 3rd Maximum exist(because set does not contain duplicate element)
            int Third_index_from_last=s.size()-3;
            auto third_maximum=next(s.begin(),Third_index_from_last);
            return *third_maximum;
        }
            return *--s.end(); // return maximum if 3rd maximum not exist
    }
};
